clang/lib/Parse/ParseDecl.cpp

```diff
@@ -8941,6 +8941,88 @@ TypeResult Parser::ParseTypeFromString(StringRef TypeStr, StringRef Context,
   return Result;
 }
 
+ExprResult
+Parser::ParseBoundsAttributeArgFromString(StringRef ExprStr, StringRef Context,
```

Comment:
I am not sure how heavy weight this is, do you think it would make sense to only spin all of these up once per declaration instead of once per annotation?  Feel free to leave as is for now but maybe we could leave a fixme or todo in case this shows up in a profile.

Reply:
I think it's tricky, because IIUC the Lexer is immutably tied to a FileID, which is tied to a read-only MemoryBuffer, so we have to construct new ones for each string. In ClangImporter we get around this by caching the parse result of e.g. the swift_attr strings. Of course, even if the string content we're parsing would be the exact same, the resulting DeclRefs would vary depending on the surrounding scope, so we'd have to flush the cache for each new declaration. I don't think we'd have enough cache hits for identical attribute arguments within a single declaration to warrant it.

Reply:
Oh, I see.